File: Tree & Binary Tree/Kth Smallest Element in a BST/solution_iterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        node = root
        while (node or stack) and k:
            if node:
                stack.append(node)
                node = node.left
            else:
                node = stack.pop()
                k -= 1
                if k == 0:
                    return node.val
                node = node.right
        return node.val
            

File: Tree & Binary Tree/Binary Tree Inorder Traversal/solution_iterative.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        if(root == nullptr) return {};
        vector<int> result;
        stack<TreeNode*> st;
        while(root != nullptr) {
            st.push(root);
            root = root->left;
        }
        while(st.size() > 0) {
            auto cur = st.top();
            st.pop();
            result.push_back(cur->val);
            cur = cur->right;
            while(cur != nullptr) {
                st.push(cur);
                cur = cur->left;
            }
        }
        return result;
    }
};
